### PR TITLE
Don't merge yet: MegaScale API Changes 2

### PR DESCRIPTION
**Pull request by jamesagnew**

### CODE REVIEW

**Review by JasonRoberts-smile — APPROVED** [review]
Approved, with comments

File: hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/BaseSearchParamExtractor.java [hunk 1]
```diff
@@ -1567,6 +1567,7 @@ public void start() {
 		myAddressCityValueChild = addressDefinition.getChildByName("city");
 		myAddressDistrictValueChild = addressDefinition.getChildByName("district");
 		myAddressStateValueChild = addressDefinition.getChildByName("state");
+		myAddressStateValueChild = addressDefinition.getChildByName("state");
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/BaseSearchParamExtractor.java:1570]: Oops - Fixed

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/entity/BulkExportJobEntity.java [hunk 2]
```diff
@@ -68,13 +70,15 @@ public class BulkExportJobEntity implements Serializable {
 
 	public static final int REQUEST_LENGTH = 1024;
 	public static final int STATUS_MESSAGE_LEN = 500;
+	public static final String JOB_ID = "JOB_ID";
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/entity/BulkExportJobEntity.java:73]: Well, FWIW it's extracted so I could refer to it from the migration tasks - Making it (at least to me) not totally useless...

File: hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirSystemDaoR4Test.java [hunk 3]
```diff
@@ -38,42 +38,14 @@
 import org.hamcrest.Matchers;
 import org.hl7.fhir.instance.model.api.IAnyResource;
 import org.hl7.fhir.instance.model.api.IIdType;
-import org.hl7.fhir.r4.model.AllergyIntolerance;
-import org.hl7.fhir.r4.model.Appointment;
-import org.hl7.fhir.r4.model.Attachment;
-import org.hl7.fhir.r4.model.Binary;
-import org.hl7.fhir.r4.model.Bundle;
+import org.hl7.fhir.r4.model.*;
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-test-r4/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirSystemDaoR4Test.java:41]: Fixed - This wasn't deliberate but I don't know what's going on. Lately IJ seems to occasionally completely ignore the editorconfig directive that bans those star imports....

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/dao/IFhirResourceDao.java [hunk 4]
```diff
@@ -206,31 +206,26 @@ default T readByPid(IResourcePersistentId thePid, boolean theDeletedOk) {
...
-	 */
-	BaseHasResource readEntity(IIdType theId, boolean theCheckForForcedId, RequestDetails theRequest);
+	IBasePersistedResource readEntity(IIdType theId, RequestDetails theRequest);
 
 	/**
 	 * Updates index tables associated with the given resource. Does not create a new
 	 * version or update the resource's update time.
 	 */
-	void reindex(T theResource, ResourceTable theEntity);
+	void reindex(T theResource, IBasePersistedResource theEntity);
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/dao/IFhirResourceDao.java:228]: Done

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/dao/IFhirResourceDao.java [hunk 5]
```diff
@@ -206,31 +206,26 @@ default T readByPid(IResourcePersistentId thePid, boolean theDeletedOk) {
...
 	T read(IIdType theId, RequestDetails theRequestDetails, boolean theDeletedOk);
 
-	BaseHasResource readEntity(IIdType theId, RequestDetails theRequest);
-
-	/**
-	 * @param theCheckForForcedId If true, this method should fail if the requested ID contains a numeric PID which exists, but is
-	 *                            obscured by a "forced ID" so should not exist as far as the outside world is concerned.
-	 */
-	BaseHasResource readEntity(IIdType theId, boolean theCheckForForcedId, RequestDetails theRequest);
+	IBasePersistedResource readEntity(IIdType theId, RequestDetails theRequest);
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/api/dao/IFhirResourceDao.java:222]: Added

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/dao/tx/HapiTransactionService.java [hunk 6]
```diff
@@ -197,6 +192,34 @@ protected <T> T doExecute(ExecutionBuilder theExecutionBuilder, TransactionCallb
 			ourRequestPartitionThreadLocal.set(requestPartitionId);
 		}
 
+		if (Objects.equals(previousRequestPartitionId, requestPartitionId)) {
+			if (TransactionSynchronizationManager.isActualTransactionActive()) {
+				if (!TransactionSynchronizationManager.isCurrentTransactionReadOnly() || theExecutionBuilder.myReadOnly) {
+					if (theExecutionBuilder.myPropagation == null || theExecutionBuilder.myPropagation == Propagation.REQUIRED) {
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/dao/tx/HapiTransactionService.java:198]: CHanged as suggested

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/util/CompositeInterceptorBroadcaster.java [hunk 7]
```diff
@@ -52,6 +52,7 @@ public static Object doCallHooksAndReturnObject(IInterceptorBroadcaster theInter
 		return newCompositeBroadcaster(theInterceptorBroadcaster, theRequestDetails).callHooksAndReturnObject(thePointcut, theParams);
 	}
 
+	// TODO: JA - Refactor to make thePointcut the last argument in order to be consistent with thr other methods here
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/util/CompositeInterceptorBroadcaster.java:55]: Future self

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/partition/BaseRequestPartitionHelperSvc.java [hunk 8]
```diff
@@ -94,19 +92,18 @@ public BaseRequestPartitionHelperSvc() {
...
 			// Handle system requests
 			//TODO GGG eventually, theRequest will not be allowed to be null here, and we will pass through SystemRequestDetails instead.
-			if ((theRequest == null || theRequest instanceof SystemRequestDetails) && nonPartitionableResource) {
-				return RequestPartitionId.defaultPartition();
-			}
-
-			if (theRequest instanceof SystemRequestDetails && systemRequestHasExplicitPartition((SystemRequestDetails) theRequest)) {
+			if (theRequest instanceof SystemRequestDetails && systemRequestHasExplicitPartition((SystemRequestDetails) theRequest) && !nonPartitionableResource) {
 				requestPartitionId = getSystemRequestPartitionId((SystemRequestDetails) theRequest, nonPartitionableResource);
+			} else if ((theRequest == null || theRequest instanceof SystemRequestDetails) && nonPartitionableResource) {
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/partition/BaseRequestPartitionHelperSvc.java:105]: Agree that's weird... have reworked this block

**Pull request merged**